### PR TITLE
Allow watching the root directory

### LINKED ISSUE #12135: Support `SUBST` or disable its use with file watching
`SUBST` is not currently supported and we fail the `SubstIntegrationTest` on Windows.

We should either add support for it or fail/degrade in a meaningful way.

---
cc: @gradle/build-cache

### PR DESCRIPTION
**Pull request by wolfs**
Watching the root directory may happen when `subst` is used to map a project directory to a drive letter.

Resolves part of #12135.

### CODE REVIEW

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchRootUtil.java [hunk 1]
```diff
@@ -79,6 +79,9 @@
             default:
                 throw new AssertionError();
         }
+        if (ancestorToWatch == null) {
+            return ImmutableList.of(path);
+        }
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchRootUtil.java:84]: We should have a comment here explaining that this is about watching the root of a file-system.

Comment (lptr) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchRootUtil.java:84]: Does this mean that if we have a typo in a file's path (e.g. like `C:\rogram Files\Something\Something`), we'll end up watching the whole drive? I think we need to be more explicit, and only check if the direct parent of the actual path is `null`.

Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchRootUtil.java:84]: We already do watch the root drive if there is a type in the file's path. This PR doesn't change anything about that. `findFirstExistingAncestor` also never returns `null` and fails when the drive is missing. I wonder if we can do something different than failing when the whole drive is missing on `Windows`. I suppose we would need to find a way to listen for drive creation events.

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchRootUtil.java (hunk 1 above)
Comment (lptr) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchRootUtil.java]: I'm not so worried about drive creation events, at least not until we have a concrete use case for where it's broken.

I'm a bit more worried about missing files potentially leading to watching the entire drive. What about this scenario:

* I'm trying to watch `C:\a\b\c\d\e`, and `C:\a` (along with the rest) does not exist, so I end up watching `C:\`
* Then `C:\a\b\c` gets created. In this case it seems we should stop watching `C:\`, and start watching `C:\a\b\c` instead. I don't think we do this right now, right?

Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchRootUtil.java]: We already do this (watch the subdirectory when it appears). I added a test.

**Pull request merged**